File: juno_rs/src/strategies/sig_osc.rs
use super::{
    Oscillator, OscillatorParams, OscillatorParamsContext, Signal, SignalParams,
    SignalParamsContext, Strategy, StrategyMeta,
};
use crate::{
    genetics::Chromosome,
    utils::{combine, MidTrend, MidTrendPolicy, MidTrendPolicyExt, Persistence},
    Advice, Candle,
};
use juno_derive_rs::*;
use rand::prelude::*;
use serde::{Deserialize, Serialize};
use std::cmp::{max, min};

#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
pub enum OscFilter {
    Enforce,
    Prevent,
}

#[derive(Chromosome, Clone, Copy, Debug, Deserialize, Serialize)]
pub struct SigOscParams {
    #[chromosome]
    pub sig: SignalParams,
    #[chromosome]
    pub osc: OscillatorParams,
    pub osc_filter: OscFilter,
    pub persistence: u32,
    pub mid_trend_policy: MidTrendPolicy,
}

fn persistence(rng: &mut StdRng) -> u32 {
    rng.gen_range(0..10)
}
fn mid_trend_policy(rng: &mut StdRng) -> MidTrendPolicy {
    rng.gen_mid_trend_policy()
}
fn osc_filter(rng: &mut StdRng) -> OscFilter {
    if rng.gen_bool(0.5) {
        OscFilter::Enforce
    } else {
        OscFilter::Prevent
    }
}

#[derive(Signal)]
pub struct SigOsc {
    sig: Box<dyn Signal>,
    osc: Box<dyn Oscillator>,
    osc_filter: OscFilter,
    advice: Advice,
    mid_trend: MidTrend,
    persistence: Persistence,
    t: u32,
    t1: u32,
}

impl SigOsc {
    pub fn new(params: &SigOscParams, meta: &StrategyMeta) -> Self {
        let sig = params.sig.construct(meta);
        let osc = params.osc.construct(meta);
        let mid_trend = MidTrend::new(params.mid_trend_policy);
        let persistence = Persistence::new(params.persistence, false);
        Self {
            advice: Advice::None,
            t: 0,
            t1: max(sig.maturity(), osc.maturity())
                + max(mid_trend.maturity(), persistence.maturity())
                - 1,
            sig,
            osc,
            osc_filter: params.osc_filter,
            mid_trend,
            persistence,
        }
    }

    fn filter(&self, advice: Advice) -> Advice {
        match self.osc_filter {
            OscFilter::Enforce => self.filter_enforce(advice),
            OscFilter::Prevent => self.filter_prevent(advice),
        }
    }

    fn filter_enforce(&self, advice: Advice) -> Advice {
        match advice {
            Advice::None => Advice::None,
            Advice::Liquidate => Advice::Liquidate,
            Advice::Long => {
                if self.osc.oversold() {
                    Advice::Long
                } else {
                    Advice::Liquidate
                }
            }
            Advice::Short => {
                if self.osc.overbought() {
                    Advice::Short
                } else {
                    Advice::Liquidate
                }
            }
        }
    }

    fn filter_prevent(&self, advice: Advice) -> Advice {
        match advice {
            Advice::None => Advice::None,
            Advice::Liquidate => Advice::Liquidate,
            Advice::Long => {
                if self.osc.overbought() {
                    Advice::Liquidate
                } else {
                    Advice::Long
                }
            }
            Advice::Short => {
                if self.osc.oversold() {
                    Advice::Liquidate
                } else {
                    Advice::Short
                }
            }
        }
    }
}

impl Strategy for SigOsc {
    fn maturity(&self) -> u32 {
        self.t1
    }

    fn mature(&self) -> bool {
        self.t >= self.t1
    }

    fn update(&mut self, candle: &Candle) {
        self.t = min(self.t + 1, self.t1);

        self.sig.update(candle);
        self.osc.update(candle);

        if self.sig.mature() && self.osc.mature() {
            let advice = self.filter(self.sig.advice());
            self.advice = combine(
                self.mid_trend.update(advice),
                self.persistence.update(advice),
            );
        }
    }
}

File: juno/strategies/sig_osc.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from juno import Advice, Candle, strategies
from juno.config import init_module_instance

from .strategy import MidTrend, MidTrendPolicy, Oscillator, Persistence, Signal


@dataclass
class SigOscParams:
    sig: Any
    osc: Any
    osc_filter: str = "enforce"
    mid_trend_policy: MidTrendPolicy = MidTrendPolicy.CURRENT
    persistence: int = 0

    def construct(self) -> SigOsc:
        return SigOsc(self)


# Generic signal with additional oscillator, persistence and mid trend filters.
#
# In order for the signal to be valid, the oscillator must be, in case of:
# - 'enforce' filter - oversold when going long, or overbought when going short
# - 'prevent' filter - not overbought when going long, and not oversold when going short
class SigOsc(Signal):
    _advice: Advice = Advice.NONE
    _sig: Signal
    _osc: Oscillator
    _osc_filter: str
    _mid_trend: MidTrend
    _persistence: Persistence
    _t: int = 0
    _t1: int

    def __init__(self, params: SigOscParams) -> None:
        assert params.osc_filter in ["enforce", "prevent"]

        self._sig = init_module_instance(strategies, params.sig)
        self._osc = init_module_instance(strategies, params.osc)
        self._osc_filter = params.osc_filter
        self._mid_trend = MidTrend(params.mid_trend_policy)
        self._persistence = Persistence(level=params.persistence, return_previous=False)
        self._t1 = (
            max(self._sig.maturity, self._osc.maturity)
            + max(self._mid_trend.maturity, self._persistence.maturity)
            - 1
        )

    @property
    def advice(self) -> Advice:
        return self._advice

    @property
    def maturity(self) -> int:
        return self._t1

    @property
    def mature(self) -> bool:
        return self._t >= self._t1

    def update(self, candle: Candle) -> None:
        self._t = min(self._t + 1, self._t1)

        self._sig.update(candle)
        self._osc.update(candle)

        if self._sig.mature and self._osc.mature:
            advice = self._sig.advice

            if self._osc_filter == "enforce":
                advice = self._osc_enforce(advice)
            else:
                advice = self._osc_prevent(advice)

            self._advice = Advice.combine(
                self._mid_trend.update(advice),
                self._persistence.update(advice),
            )

    def _osc_enforce(self, advice: Advice) -> Advice:
        return (
            Advice.LIQUIDATE
            if (
                advice is Advice.LONG
                and not self._osc.oversold
                or advice is Advice.SHORT
                and not self._osc.overbought
            )
            else advice
        )

    def _osc_prevent(self, advice: Advice) -> Advice:
        return (
            Advice.LIQUIDATE
            if (
                advice is Advice.LONG
                and self._osc.overbought
                or advice is Advice.SHORT
                and self._osc.oversold
            )
            else advice
        )
